@here/harp-mapview/lib/ClipPlanesEvaluator.ts

```diff
@@ -0,0 +1,524 @@
+ * and multiplier.
+ */
+export class InterpolatedClipPlanesEvaluator implements ClipPlanesEvaluator {
+    readonly farMin: number;
+
+    protected m_tmpVectors: THREE.Vector3[] = [
+        new THREE.Vector3(),
+        new THREE.Vector3(),
+        new THREE.Vector3()
+    ];
+    protected m_tmpQuaternion: THREE.Quaternion = new THREE.Quaternion();
+
+    constructor(
+        readonly nearMin: number = 0.1,
+        readonly nearMultiplier: number = 0.1,
+        readonly nearFarMultiplier = 50.0,
+        readonly farOffset = 200.0
+    ) {
+        assert(nearMin > 0);
+        assert(nearFarMultiplier >= 0);
+        assert(farOffset >= 0);
+        this.farMin = nearMin * nearFarMultiplier + farOffset;
+    }
+
+    evaluateClipPlanes(
+        camera: THREE.Camera,
+        projection: Projection
+    ): { near: number; far: number } {
+        let nearPlane: number = this.nearMin;
+        let farPlane: number = this.farMin;
+        if (projection.type === ProjectionType.Spherical) {
+            // near and far plane for a set up where
+            // the camera is looking at the center of the scene.
+            const r = EarthConstants.EQUATORIAL_RADIUS;
+            const d = camera.position.length();
+            const alpha = Math.asin(r / d);
+            // Extract X, Y, Z axes into tmp vectors array.
+            camera.matrixWorld.extractBasis(
+                this.m_tmpVectors[0],
+                this.m_tmpVectors[1],
+                this.m_tmpVectors[2]
+            );
+            // Setup quaternion based on X axis.
+            this.m_tmpQuaternion.setFromAxisAngle(this.m_tmpVectors[0], alpha);
+            // Acquire forward vector based on Z axis reversed (keep it in tmpVectors[2]).
+            const fwd = this.m_tmpVectors[2].negate();
+            // Apply quaternion rotation to forward vector, store it in tmpVectors[1].
+            const fwdRot = this.m_tmpVectors[1].copy(fwd).applyQuaternion(this.m_tmpQuaternion);
+            // Store camera position tmpVectors[0] and reference it with p.
+            const p = this.m_tmpVectors[0].copy(camera.position);
+            p.addScaledVector(fwdRot, Math.sqrt(d * d - r * r));
+            farPlane = p.sub(camera.position).dot(fwd);
+            const bias = 2000; // TODO: generalize.
+            nearPlane = Math.max(this.nearMin, projection.groundDistance(camera.position) - bias);
+        } else {
+            const groundDistance = projection.groundDistance(camera.position);
+            nearPlane = Math.max(this.nearMin, groundDistance * this.nearMultiplier);
+            // Will be already clamped to minFar due to clamping above.
+            farPlane = nearPlane * this.nearFarMultiplier + this.farOffset;
+        }
+        return { near: nearPlane, far: farPlane };
+    }
+}
+
+/**
+ * Simple camera clip planes evaluator that computes near plance based on camera to ground distance.
```

Comment:
From reading the documentation, I don't understand what difference this has with the above InterpolatedClipPlanesEvaluator. Perhaps describe which use cases suit which implementation.

Reply:
Will do so. Generally do not bother about InterpolatedClipPlanesEvaluator, it is kept just to have all examples in tact. The implementation of this class comes from legacy (original) code that was used, its very imprecise and overgrows the camera frustum. My idea is to remove it totally. As you probably saw there are also some hard-code values in spherical projection case, but this is how it was originally. I did not wanted to make total revolution here, because a lot of stuff depends on that. So its quite big topic anyway and I feel that requires gentle (step-by-step) improvements.